### PR TITLE
feat: add blockSpecifer for blockParameter to support Eip1898

### PR DESCRIPTION
**Pull request by RetricSu**
```sh
{
    "jsonrpc": "2.0",
    "method": "eth_call",
    "params": [
        {
            "from": "0x966b30e576a4d6731996748b48dd67c94ef29067",
            "to": "0x45b2585ee6877b36e359603eae3c465392b57d13",
            "data": "0xb67d77c500000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000006"
        },
        {
            "blockHash": "0x65b3f50d57e73112852d926f5f18cd5714a487c81ed9da1b9a1a9bd16a9e2d81",
            "requireCanonical": false
        }
    ],
    "id": 1
}
```

---

```sh
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": "0x0000000000000000000000000000000000000000000000000000000000000002"
}
```

## References
- https://eips.ethereum.org/EIPS/eip-1898
- https://github.com/ethereum/go-ethereum/pull/19491

## Test
- https://github.com/nervosnetwork/godwoken-tests/pull/166

### CODE REVIEW

**Review by keroro520 — COMMENTED** [review]
Please add some test cases (https://eips.ethereum.org/EIPS/eip-1898 has listsed test cases).

File: packages/api-server/src/methods/types.ts [hunk 1]
```diff
@@ -16,7 +16,16 @@ export type Response = number | string | boolean | SyningStatus | Array<string>;
 
 export type Callback = (err: Error, res?: any | Response) => void;
 
-export type BlockParameter = HexNumber | "latest" | "earliest" | "pending";
+export type BlockTag = "latest" | "earliest" | "pending";
+
+// Eip1898 support block hash and block number
+export interface BlockSpecifier {
```
Comment (keroro520) [inline packages/api-server/src/methods/types.ts:22]: What about renaming EIP1898BlockSpecifier ?

File: packages/api-server/src/methods/validator.ts [hunk 2]
```diff
@@ -273,11 +273,62 @@ export function verifyBlockParameter(
...
   }
 
+  if (blockSpecifier.blockHash != null && blockSpecifier.blockNumber != null) {
+    return invalidParamsError(
+      index,
+      "blockHash and blockNumber can not exits at same time"
+    );
+  }
+
+  if (blockSpecifier.requireCanonical === true) {
```
Comment (keroro520) [inline packages/api-server/src/methods/validator.ts:313]: Why true is disallowed?

File: packages/api-server/src/methods/modules/eth.ts [hunk 3]
```diff
@@ -1114,8 +1114,34 @@ export class Eth {
...
+    if (
+      typeof blockParameter === "object" &&
+      blockParameter.blockHash != null
+    ) {
+      const block = await this.getBlockByHash([
+        blockParameter.blockHash,
+        false,
+      ]);
+      if (block == null) {
+        throw new InvalidParamsError(
```
Comment (keroro520) [inline packages/api-server/src/methods/modules/eth.ts:1127]: Returns HeaderNotFound?

File: packages/api-server/src/methods/types.ts (hunk 1 above)
Comment (RetricSu) [reply packages/api-server/src/methods/types.ts:22]: a bit too long. I personally think `BlockSpecifier` with commenting eip1898 is clear enough?

File: packages/api-server/src/methods/validator.ts (hunk 2 above)
Comment (RetricSu) [reply packages/api-server/src/methods/validator.ts:313]: we don't have canonical blocks, so disallowed.

to be compatible with Ethereum,  requireCanonical should be kept according to the specs of eip1898. Throw error when un-compatible is more clear for users.

File: packages/api-server/src/methods/modules/eth.ts (hunk 3 above)
Comment (RetricSu) [reply packages/api-server/src/methods/modules/eth.ts:1127]: I think `HeaderNotFound == tipBlockNumber < user request's block number` ?

File: packages/api-server/src/methods/validator.ts (hunk 2 above)
Comment (keroro520) [reply packages/api-server/src/methods/validator.ts:313]: Oh, this is the opposite of what I understood. As far as I know, Godwoken only has canonical blocks.

File: packages/api-server/src/methods/modules/eth.ts (hunk 3 above)
Comment (keroro520) [reply packages/api-server/src/methods/modules/eth.ts:1127]: > I think `HeaderNotFound == tipBlockNumber < user request's block number` ?

This meaning cannot be derived from "HeaderNotFound".

"xxxNotFound" indicates that we cannot find the corresponding resource from database.

File: packages/api-server/src/methods/validator.ts (hunk 2 above)
Comment (RetricSu) [reply packages/api-server/src/methods/validator.ts:313]: you are right,  I misunderstood it as uncle block

Comment (RetricSu) [reply packages/api-server/src/methods/validator.ts]: updated

File: packages/api-server/src/methods/modules/eth.ts (hunk 3 above)
Comment (RetricSu) [reply packages/api-server/src/methods/modules/eth.ts]: updated

**Review by keroro520 — APPROVED** [review]

**Review by Flouse — APPROVED** [review]

**Pull request merged**